CMakeLists.txt

```diff
@@ -21,9 +22,11 @@ set(CMAKE_EXPORT_COMPILE_COMMANDS ON)
 # etc.), by default some components (e.g., cmake) install into lib while
 # others (e.g., python) install into lib64.
 # This causes trouble when we try to figure out the runtime directory in
-# CompilerUtils.cpp::getLibraryPath(). So we explicitly set CMAKE_INSTALL_LIBDIR
+# CompilerUtils.cpp::getLibraryPath(). So we set a default CMAKE_INSTALL_LIBDIR
 # to install into lib on all systems so we don't have to deal with lib64.
-set(CMAKE_INSTALL_LIBDIR lib)
+if(NOT CMAKE_INSTALL_LIBDIR)
+  set(CMAKE_INSTALL_LIBDIR lib)
+endif()
```

Comment:
> We specifically restrict the LIBDIR to be "lib" because that's the only place `getLibraryPath()` function in

* And with this proposal here **will remain as** ```lib```, but user **may** and **can** override.
* Distros usually like the ```lib64``` flavour so one may override this to complain with other requirements.

> `CompilerOptions.cpp` will search. If you change it to something else like "lib64", `getLibraryPath()` must also be changed accordingly. Otherwise, it breaks the `onnx-mlir` command.

* Thanks for highlighting this, it is now propagated down  [from here](https://github.com/onnx/onnx-mlir/pull/2700/files#diff-8c9999420c6434188ef79512be385f75efd438e2ebc0d50665731a3a5e694ed8R60-R61) to [CompilerOptions.{hpp,cpp}](https://github.com/onnx/onnx-mlir/pull/2700/files#diff-125357d108ffb6b9c2d0f7baf187ae0246c75f71c71cb3f7eca72a8ffae70bc0R34-R44)

```
$ make OMCompilerOptions VERBOSE=1

Building CXX object src/Compiler/CMakeFiles/OMCompilerOptions.dir/CompilerOptions.cpp.o
/usr/bin/clang++ -DLIBDIR_PATH=\"lib64\"  {...}
```

Reply:
* Agree, but it cost nothing for now to have an option to control this.
* There is now a single central place to configure this, in the very ```cmake```.

Also in the future, it can be dropped at all, we can have a smarter way to lookup via ```libc``` in all places where liobraries lives including the standard extra ```LD_LIBRARY_PATH```.

Reply:
What do you mean it costs nothing? You just added code in `CompilerOptions.{h|c}pp` and added string substitution with cmake target definition when we already have a different string substitution mechanism (via `ExternalUtil.hpp.in`). So now we have to manage two different ways to do string substitution. Allowing a different LIBDIR doesn't bring enough value to justify these added complexity. It was also our intention to have the same LIBDIR on all distros to make things simpler. We do that also for dependents like protobuf, i.e., we install protobuf under "lib" on all distros.

Reply:
> What do you mean it costs nothing? You just added code in `CompilerOptions.{h|c}pp` and added string substitution with cmake target definition when we already have a different string substitution mechanism (via `ExternalUtil.hpp.in`). So now we have to manage two different ways to do string substitution. 

* Where exactly in `ExternalUtil.hpp.in` ? If so can be expanded from the central ```CMAKE_INSTALL_LIBDIR```.

Allowing a different LIBDIR doesn't bring enough value to justify these added complexity. It was also our intention to have the same LIBDIR on all distros to make things simpler. We do that also for dependents like protobuf, i.e., we install protobuf under "lib" on all distros.

* Here are the pros that might justify:

1. The ```llvm-project``` monorepo have the official [LLVM_LIBDIR_SUFFIX](https://github.com/llvm/llvm-project/blob/5bcd91058ee4855804780c4ae35ac87ed45a4b58/llvm/docs/CMake.rst#L247) exactly for this:
    * See sample usage in RHEL case when building the llvm-project monorepo:
      ```
      %if 0%{?__isa_bits} == 64
         -DLLVM_LIBDIR_SUFFIX=64 \
      %else
        -DLLVM_LIBDIR_SUFFIX= \
      %endif
      ```
   * Would you like to use provided ```LLVM_LIBDIR_SUFFIX``` instead to controll this ?


  2. SLES/RHEL still use "lib64" and this will not go away soon.

Reply:
It doesn't matter what llvm-project did. Every project has its own logic and we consciously made the decision to install under "lib" on all distros to simplify code and build scripts. Installing under "lib64" on RHEL/SLES brings no visible value to us to justify the cost other than the fact that some other packages are still using "lib64". But even on RHEL/SLES, nobody says you have to install under "lib64". For example, pip3 on RHEL/SLES, java on RHEL, etc. install under "/usr/lib" instead of "/usr/lib64". In fact, the primary reason for using "lib64" should be that if you support both 32-bit and 64-bit libraries with the same name. Otherwise, there is no reason to use "lib64" at all.

Reply:
@gongsu832 

> It doesn't matter what llvm-project did. Every project has its own logic and we consciously made the decision to install under "lib" on all distros to simplify code and build scripts. 

* These libraries are supposed to be community friendly, offering a minimal interoperability and user friendliness. I am quite sure in case of llvm-project, a very active one, the conscious decisions where made to the benefit of a broad community.

Thanks for you time.